Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,200 @@
+be too complex for a single device to be managed as a monolithic block or
+a part of the functionality might need to be exposed to a different subsystem.
+Splitting the functionality into smaller orthogonal devices would make it
+easier to manage data, power management and domain-specific communication with
+the hardware. A key requirement for such a split is that there is no dependency
+on a physical bus, device, register accesses or regmap support. These
+individual devices split from the core cannot live on the platform bus as they
+are not physical devices that are controlled by DT/ACPI. The same argument
+applies for not using MFD in this scenario as it relies on individual function
+devices being physical devices that are DT enumerated.
+
+An example for this kind of requirement is the audio subsystem where a single
+IP handles multiple entities such as HDMI, Soundwire, local devices such as
+mics/speakers etc. The split for the core's functionality can be arbitrary or
+be defined by the DSP firmware topology and include hooks for test/debug. This
+allows for the audio core device to be minimal and tightly coupled with
+handling the hardware-specific logic and communication.
+
+The ancillary bus is intended to be minimal, generic and avoid domain-specific
+assumptions. Each ancillary_device represents a part of its parent
+functionality. The generic behavior can be extended and specialized as needed
+by encapsulating an ancillary_device within other domain-specific structures and
+the use of .ops callbacks. Devices on the same ancillary bus do not share any
+structures and the use of a communication channel with the parent is
+domain-specific.
+
+When Should the Ancillary Bus Be Used
+=====================================
+
+When a kernel driver registers either an ancillary_device or ancillary_driver on
+the ancillary bus, we can use the nomenclature to refer to this kernel driver as
+a registering driver.  It is the entity that will allocate memory for and
+register something on the ancillary bus.
+
+The ancillary bus is to be used when multiple registering drivers, who share a
+common header file, need a mechanism to make connection and provide access to a
+shared struct allocated by the registering driver(s) that registered the
+ancillary_device(s).  Each of these registering drivers can be from the same
+subsystem, or from multiple subsystems.
+
+The emphasis here is on a common generic interface that keeps subsystem
+customization out of the bus infrastructure.
+
+When Should the Ancillary Bus Not Be Used
+=========================================
+
+If there is a need for a heavier weight bus that provides functionality beyond
+just the matching up of multiple registering drivers, and sacrifices common
+generality to extend the bus functionality into subsystem specific functions,
+then the ancillary bus should not be used.  A custom bus solution, or some other
+method should be considered in this use case.
```

Comment:
@ranj063 we discussed with Greg KH that virtbus will be used beyond just the matching service instead of creating one.
It is beyond just the matching service, did you discuss this over email to limit the usage when Greg KH agreed last time?

Reply:
@paravmellanox could you please elaborate on what you mean on "beyond just the matching service". Put differently, what would be missing from the current implementation of the ancillary bus for your needs?

Reply:
but to be clear Greg KH was informed in the email thread for the previous series.

Reply:
@ranj063 
Please follow through the thread and discussion [1] with Greg KH.
We must converge before posting this series and have good enough answer to prove Greg KH suggestions wrong. I am not sure you want to go that path at this point.

[1] https://patchwork.kernel.org/patch/11280547/#23056771

Reply:
I looked through the linked thread and I didn't see any mention of any functionality beyond the matching.  The primary focus of ancillary bus is to provide a way for multiple kernel objects to be matched and have access to a shared object.  Emphasis on generic interface and commonality.

If I missed something could you please point me to it?

Thanks

Reply:
in [2] refer to this line.
cat /sys/bus/virtbus/devices/mlx5_sf.1/sfnum

Ignore the incorrect device name mlx5_sf.1, it is virtdev.1.
But the part to focus is additional sysfs attributes that will be used by systemd to rename the netdevices of this virtdev (sub function).

At beginning there is only handful attribute like sfnum. But as it becomes mature, it will have more attributes like irqs, numa hierarchy of device, because device should be self-contained for such device specific properties.

It has functionality beyond matching, that is: a sub-function device is anchored on the virtbus. There will be netdevice and rdma device attached to this device.
So when someone sees the sysfs files as below there are actual net and rdma devices attached to it.

/sys/bus/virtbus/devices/virtdev.1/net/eth0
/sys/bus/virtbus/devices/virtdev.1/infiniband/mlx5_2

When someone uses the bus as matching service bus, there won't be class devices underneath like above.

This subtle difference and information is capture and acked in thread by Greg KH in [1].

[1] https://patchwork.kernel.org/patch/11280547/#23056985
[2] https://lore.kernel.org/netdev/20200519092258.GF4655@nanopsycho/